### ISSUE TITLE
Support concurrent metadata fetching

### ISSUE THREAD

**Issue by tdejager** [mamba-org/resolvo#3]
This explains it well: https://github.com/pubgrub-rs/pubgrub/issues/138

Also see: https://github.com/orogene/orogene/blob/main/crates/node-maintainer/src/resolver.rs where this is implemented in a tree resolver.

I think we could either make the `DependencyProvider` async. Or provide a sync and async version or something.

Next to that we should either make `resolvo` entirely async or use something like a queue to wait on metadata fetching and use this information once it comes available, like I said orogone takes this approach.

**Comment by aochagavia** [comment]
I've been thinking about async and here is my current take:

1. Resolvo is heavily CPU intensive, which usually does not work well with async systems. IMO it makes more sense for it to be blocking, at least at the API level.
2. Internally, we _could_ use async for network requests. I'll think about that :)

**Comment by tdejager** [comment]
So those functions can be used in an asynchronous context. Internally we could use a thread pool to dispatch those tasks for example. But otherwise asynchronous code like RIP needs to use the 'block_on' functions to be able to implement the DependencyProvider

**Comment by aochagavia** [comment]
Ah, now I understand better, thanks!

**Comment by Eh2406** [comment]
There is always the risk that the main loop will get stuck so long (doing blocking work) that the asynchronous calls are no longer valid when the library eventually yields. (TCP requires handshakes to be responded to promptly or connections get closed.) Mixing blocking code like the resolution loop and non blocking asynchronous call is tricky. If you find a good solution, let me know and I will copy it in pubgrub.

**Comment by tdejager** [comment]
Yes your implementation sounds like I had in mind, not sure if we can use the newly stabilized asynchronous trait or if it is too limited still.

Thank you for the comment though :)

**Comment by tdejager** [comment]
Orogone has implemented something similar: https://github.com/orogene/orogene/blob/main/crates/node-maintainer/src/resolver.rs

**Comment by baszalmstra** [comment]
this has been implemented a while ago
